### ISSUE TITLE
Ambiguity in Pool spec (AveragePool and MaxPool)

### ISSUE THREAD

**Issue by hariharans29** [onnx/onnx#1922]
Hello,

We have a couple of questions regarding the spec of AveragePool and MaxPool when a value in the output tensor is generated by a "pure padding region"

1) AveragePool: input = [[[1, 2, 3, 4, 5]]], kernel_shape = [2], pads = [2, 3], strides = [2], count_include_pad = 0, in this case, the first output value in the output tensor has to be generated by "pure padding region", and since count_include_pad = 0, should the output be nan (division by 0) or 0 (since it did not see any non-padded input) ? I suggest we modify the spec to make the value 0 to reduce ambiguity.

2) MaxPool:  For the same example as above, if the first output value is generated by "a pure padding zone", what should the value be? Additionally in the "indices" tensor, what will the first value be? It cannot be in [0, N *C * D1 *... *Dn). I suggest it be a negative value.

What do you guys think ? @houseroad @zrphercule @pranavsharma @tracysh

Thanks!

**Comment by divyanshugit** [comment]
If no one is working on this, then I would like to work on this. @jcwchen could you please give me some pointers to get started with it?

**Comment by jcwchen** [comment]
Thank you for the interest. You will need to resolve the ambiguity and update files accordingly. https://github.com/onnx/onnx/pull/3706 This is an example PR. You can update the C++ file (onnx/defs/xxx/defs.cc) and then use `python onnx/defs/gen_doc.py` to generated new documents (docs/Changelog.md and docs/Operators.md)

**Comment by divyanshugit** [comment]
Implementation of AveragePool across PyTorch, TensorFlow, and Onnx.

PyTorch
----------
```python

import torch

avg_pool = torch.nn.AvgPool2d((3, 3), stride=(3,4),ceil_mode=True,count_include_pad=False)
input = torch.ones((1,1,8,8))

output = avg_pool(input)

print(output)
>>> tensor([[[[1., 1.],
          [1., 1.],
          [1., 1.]]]])

print(output.shape)
>>> torch.Size([1, 1, 3, 2])

```
TensorFlow
----------------
```python

import tensorflow as tf

avg_pool = tf.keras.layers.AveragePooling2D(pool_size=(3, 3),
   strides=(3, 4), padding='same')
x = tf.ones((1,8,8,1))

output = avg_pool(x)

print(output)
>>> tf.Tensor(
[[[[1.]
   [1.]]

  [[1.]
   [1.]]

  [[1.]
   [1.]]]], shape=(1, 3, 2, 1), dtype=float32)

print(output.shape)
>>> (1,3,2,1)

```
As we can see above; in TensorFlow and PyTorch the calculation of AveragePool is as([Source](https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html)):
```math
        out(N_i, C_j, h, w)  = \frac{1}{kH * kW} \sum_{m=0}^{kH-1} \sum_{n=0}^{kW-1}
                               input(N_i, C_j, stride[0] \times h + m, stride[1] \times w + n)

```
But in the case of onnx; Ig, it adds an extra dimension to incorporate the difference in the representation of `Channel * Height * Width`. In torch, an image is represented by `Channel * Height * Width` whereas in TensorFlow ` Height * Width * Channel`


Onnx
-----
```python

import onnx
import onnxruntime as ort
import numpy as np

# define the input data
data = np.ones((1, 1, 8, 8)).astype(np.float32)

# define the pooling parameters
kernel = (3, 3)
stride = (3, 4)
pad = (0, 0, 0, 0)
ceil_mode = 1
count_include_pad = False

# create the ONNX model
node = onnx.helper.make_node('AveragePool',
                             inputs=['data'],
                             outputs=['output'],
                             kernel_shape=kernel,
                             strides=stride,
                             pads=pad,
                             ceil_mode=ceil_mode,
                             count_include_pad=count_include_pad)
graph = onnx.helper.make_graph([node],
                               'test',
                               inputs=[onnx.helper.make_tensor_value_info('data', onnx.TensorProto.FLOAT, [1, 1, 8, 8])],
                               outputs=[onnx.helper.make_tensor_value_info('output', onnx.TensorProto.FLOAT, [1, 1, 2, 2])])
model = onnx.helper.make_model(graph)

# create the ONNX Runtime session and run the model
sess = ort.InferenceSession(model.SerializeToString())
input_data = {'data': data}
output = sess.run(None, input_data)


print(output)
>>> [array([[[[ 1.,  1., nan],
         [ 1.,  1., nan],
         [ 1.,  1., nan]]]], dtype=float32)]
print(output[0].shape)
>>> (1, 1, 3, 3)

```

**Comment by divyanshugit** [comment]
So, from my perspective; It'll be best to add a warning in the description of AveragePool. To resolve the issue;
Like dimension of stride should not be more than the dimension of the kernel when the `count_include_false=False`. Let me know your thoughts @jcwchen

**Comment by jcwchen** [comment]
Thanks @divyanshugit for the investigation. To be consistent with PyTorch/TensorFlow, onnx will need to change the output shape, which will potentially risk breaking change. It's better for us to choose onnx should perform NAN or 0 in this case, with the same shape ONNX has right now, and then specify the decided behavior in the doc to prevent confusion. cc @onnx/sig-operators-approvers for more inputs. Thanks!

**Comment by titaiwangms** [comment]
https://github.com/microsoft/onnxruntime/pull/16752

**Comment by justinchuby** [comment]
https://github.com/onnx/onnx/pull/5741

**Comment by justinchuby** [comment]
@titaiwangms Do you have a suggestion on how we should handle this issue?

**Comment by sfatimar** [comment]
It does not seem like this would make it in by 1.23
